pure-pulsars/src/wikiutils.py

```diff
@@ -0,0 +1,530 @@
+        if query.casefold().strip() == result.title().casefold().strip():
+            return result
+
+    # Check for near match in the title
+    for result in results:
+        if query.casefold().strip() in result.title().casefold().strip():
+            return result
+
+    return None
+
+
+async def search_wikipedia(query: str) -> Page | None:
+    """Search wikipedia and return the first result.
+
+    Args:
+    ----
+    query (str): The query to search for.
+
+    Returns:
+    -------
+    Page: The first page found. None if no results are found.
+
+    """
+    result = Page(site, title=query)
+
+    if not result.exists() or result.isRedirectPage():
+        # Try performing a search.
+        results = tuple(site.search(query, total=10))
+
+        if not results:
+            return None
+
+        return get_best_result(results, query)
+
+    try:
+        return result
+
+    except IndexError:
+        logging.exception("No results found for query: %s", query)
+        return None
+
+
+async def search_wikipedia_generator(query: str, max_number: int = 10) -> AsyncGenerator[Page, None]:
+    """Search wikipedia and return results matching the query.
+
+    Args:
+    ----
+    query (str): The query to search for.
+    max_number (int): The maximum number of results to return.
+
+    Returns:
+    -------
+    Page: The first page found. None if no results are found.
+
+    """
+    results = tuple(site.search(query, total=max_number))
+
+    for result in results:
+        yield result
+
+
+async def rand_wiki() -> Page:
+    """Return a random popular wikipedia article."""
+    return await ArticleGenerator().fetch_article()
+
+
+async def loss_update(guild: int, user: User) -> None:
+    """Update the user in the database.
+
+    Args:
+    ----
+    guild (int): The guild id.
+    user (User): The user to update.
+    score (int): The score to update.
+
+    """
+    uid = user.id
+    try:
+        db_ref_user = await DATA.get_user(guild, uid)
+        await DATA.update_value_for_user(
+            guild_id=guild,
+            user_id=uid,
+            key="times_played",
+            value=db_ref_user["times_played"] + 1,
+        )
+
+        await DATA.update_value_for_user(
+            guild_id=guild,
+            user_id=uid,
+            key="failure",
+            value=db_ref_user["failure"] + 1,
+        )
+
+        await DATA.update_value_for_user(
+            guild_id=guild,
+            user_id=uid,
+            key="last_played",
+            value=datetime.now(UTC).timestamp(),
+        )
+
+    except NullUserError:
+        new_user = UserController(
+            info=_User(
+                name=user.global_name,
+                times_played=1,
+                failure=1,
+                score=0,
+                last_played=datetime.now(UTC).timestamp(),
+            ),
+        )
+        await DATA.add_user(uid, new_user, guild)
+
+
+async def win_update(guild: int, user: User, score: int) -> None:
```

Comment:
Would it be better to have a single `update` function which you can the pass in in the win/loss argument instead of having two functions?